crates/astria-conductor/src/conductor.rs

```diff
@@ -10,122 +18,239 @@ use tokio::{
-                        Err(e) => error!(error.msg = %e, error.cause = ?e, "executor task failed"),
+                    if let Some(data_availability_reader) = data_availability_reader.take() {
+                        info!("starting data availability reader");
+                        tasks.spawn(
+                            Self::DATA_AVAILABILITY,
+                            data_availability_reader.run_until_stopped(),
+                        );
                     }
-                    break;
                 }
 
-
-                res = async { reader.as_mut().unwrap().await }, if reader.is_some() => {
+                Some((name, res)) = tasks.join_next() => {
                     match res {
-                        Ok(Ok(())) => error!("reader task exited unexpectedly"),
-                        Ok(Err(e)) => error!(error.msg = %e, error.cause = ?e, "reader exited with error"),
-                        Err(e) => error!(error.msg = %e, error.cause = ?e, "reader task failed"),
+                        Ok(Ok(())) => error!(tak.name = name, "task exited unexpectedly, shutting down"),
+                        Ok(Err(e)) => error!(task.name = name, error.msg = %e, error.cause = ?e, "task exited with error; shutting down"),
+                        Err(e) => error!(task.name = name, error.msg = %e, error.cause = ?e, "task failed; shutting down"),
                     }
-                    break;
                 }
             }
         }
 
+        info!("sending shutdown command to all tasks");
+        for (_, channel) in shutdown_channels {
+            let _ = channel.send(());
+        }
+
+        sequencer_client_pool.close();
+
+        // wait 5 seconds for all tasks to shut down
+        // put the tasks into an Rc to make them 'static
+        let mut tasks = Rc::new(tasks);
+        let local_set = LocalSet::new();
+        local_set
+            .run_until(async {
+                let mut tasks = tasks.clone();
+                let _ = timeout(
+                    Duration::from_secs(5),
+                    spawn_local(async move {
+                        while Rc::get_mut(&mut tasks)
+                            .expect(
+                                "only one Rc to the conductor tasks should exist; this is a bug",
+                            )
+                            .join_next()
+                            .await
+                            .is_some()
+                        {}
+                    }),
+                )
+                .await;
+            })
+            .await;
+
+        if !tasks.is_empty() {
+            warn!(
+                number = tasks.len(),
+                "aborting tasks that haven't shutdown yet"
+            );
+            Rc::get_mut(&mut tasks)
+                .expect("only one Rc to the conductor tasks should exist; this is a bug")
+                .shutdown()
+                .await;
+        }
```

Comment:
what's the purpose of the `Rc` and this section over just waiting 5 seconds?

Reply:
Rc: see at the top of the block where it's defined. It's to make the future 'static.

The 5 seconds are to let the tasks wind down and exit before being killed. Just a number to be honest. We should probably fine-tune it.